gapis/api/vulkan/frame_loop.go

```diff
@@ -428,18 +500,108 @@ func (f *frameLoop) resetImages(ctx context.Context, sb *stateBuilder) error {
 	for dst, src := range f.imageToBackup {
 		dstObj := s.Images().Get(dst)
 
-		primeable, err := imgPrimer.newPrimeableImageDataFromDevice(src, dst)
-		if err != nil {
-			return log.Errf(ctx, err, "Create primeable image data for image %v", dst)
+		prime := func() error {
+			primeable, err := imgPrimer.newPrimeableImageDataFromDevice(src, dst)
+			if err != nil {
+				return log.Errf(ctx, err, "Create primeable image data for image %v", dst)
+			}
+			defer primeable.free(sb)
+			err = primeable.prime(sb, useSpecifiedLayout(dstObj.Info().InitialLayout()), sameLayoutsOfImage(dstObj))
+			if err != nil {
+				return log.Errf(ctx, err, "Priming image %v with data", dst)
+			}
+			log.D(ctx, "Prime image from [%v] to [%v] succeed", src, dst)
+			return nil
 		}
-		defer primeable.free(sb)
-		err = primeable.prime(sb, useSpecifiedLayout(dstObj.Info().InitialLayout()), sameLayoutsOfImage(dstObj))
-		if err != nil {
-			return log.Errf(ctx, err, "Priming image %v with data", dst)
+
+		if err := prime(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *frameLoop) resetFences(ctx context.Context, sb *stateBuilder) error {
+	s := GetState(sb.newState)
+
+	for k := range f.fenceChanged {
+		fence := s.Fences().Get(k)
+		if fence.Signaled() {
+			log.D(ctx, "Reset fence %v.", k)
+			sb.write(sb.cb.VkResetFences(fence.Device(), 1, sb.MustAllocReadData(fence.VulkanHandle()).Ptr(), VkResult_VK_SUCCESS))
+		} else {
+			log.D(ctx, "Singal fence %v.", k)
+			queue := sb.getQueueFor(
+				VkQueueFlagBits_VK_QUEUE_GRAPHICS_BIT|VkQueueFlagBits_VK_QUEUE_COMPUTE_BIT|VkQueueFlagBits_VK_QUEUE_TRANSFER_BIT,
+				[]uint32{},
+				fence.Device(),
+				NilQueueObjectʳ)
+			if queue == NilQueueObjectʳ {
+				return log.Err(ctx, nil, "queue is nil queue")
+			}
+			sb.write(sb.cb.VkQueueSubmit(
+				queue.VulkanHandle(),
+				0,
+				memory.Nullptr,
+				fence.VulkanHandle(),
+				VkResult_VK_SUCCESS,
+			))
+
+			sb.write(sb.cb.VkQueueWaitIdle(queue.VulkanHandle(), VkResult_VK_SUCCESS))
+		}
+	}
+	return nil
+}
+
+func (f *frameLoop) resetEvents(ctx context.Context, sb *stateBuilder) error {
+	s := GetState(sb.newState)
+
+	for k := range f.eventChanged {
+		event := s.Events().Get(k)
+		if event.Signaled() {
+			sb.write(sb.cb.VkResetEvent(event.Device(), event.VulkanHandle(), VkResult_VK_SUCCESS))
+			log.D(ctx, "Reset event %v ", k)
+		} else {
+			sb.write(sb.cb.VkSetEvent(event.Device(), event.VulkanHandle(), VkResult_VK_SUCCESS))
+			log.D(ctx, "Set event %v ", k)
 		}
-		log.D(ctx, "Prime image from [%v] to [%v] succeed", src, dst)
 	}
+	return nil
+}
 
+func (f *frameLoop) resetSemaphores(ctx context.Context, sb *stateBuilder) error {
+	s := GetState(sb.newState)
+
+	for k := range f.semaphoreChanged {
```

Comment:
Similar to the others, we have to wait for the semaphore if it has been "signaled" since the signal is not complete until someone actually waits.